### DISCUSSION TITLE
为「默认播放速度」提供黑名单功能

### DISCUSSION THREAD

**Discussion by LonelySteve** [the1812/Bilibili-Evolved#1315] [功能建议 / Ideas]
大多数情况下，默认播放速度调高一点会很不错，但是如果是用来听音乐，听鬼畜，就不太合适了，每次都得手动重设回默认速度，而且播放器里的倍数状态和脚本的设定的倍数并不同步，每次还得点一下别的倍数再调回 1.0x

建议对于视频页添加一个附加功能的按钮（只在启用了「默认播放速度」这个功能时显示），可以用于切换对此视频的默认倍数的应用状态（也就是将此视频设置到「默认播放速度」的黑名单中，或从黑名单中移除）。

**Comment by the1812** [comment]
> 每次还得点一下别的倍数再调回 1.0x

这个已经在计划修复了

----

黑名单你可以弄个 PR 玩玩, 做起来应该也不难, 不过这个是细分到分P还是整个视频 (或者说存 aid 还是 cid)?

**Reply by LonelySteve** [reply]
一般来说是整个视频，如果细分到分 P 有点难受，要想想对于一些音乐合集视频可能有很多分 p，总不能一个一个地标记

**Reply by LonelySteve** [reply]
@the1812 奇怪了，之前不是有 wiki 吗？开 discussions 的时候手滑关了？😂

**Reply by the1812** [reply]
好像关了 因为里面的内容基本都过时了

**Reply by the1812** [reply]
你可以参考 src/live/live-pip.ts 里的写法, 激活条件直接用全局的 `videoCondition` 就行了

(突然想起来 [CONTRIBUTING](https://github.com/the1812/Bilibili-Evolved/blob/master/CONTRIBUTING.md) 里也有写附加功能的添加

**Reply by Geeyun-JY3** [reply]
> @the1812 奇怪了，之前不是有 wiki 吗？开 discussions 的时候手滑关了？😂

关了有一段时间了（README 也没改

**Reply by the1812** [reply]
噢 看来得清理下 wiki 链接了

**Comment by LonelySteve** [comment]
~~读了一晚上的 B 站播放器的源码，被混淆的 js 读起来真的难受~~

<details>
<summary>做梦中...</summary>

![image](https://user-images.githubusercontent.com/34429322/102237766-f61c6180-3f2f-11eb-9f04-2f75ffe9e43b.png)

如果能拿到图中的这个 `r` 对象，咱就可以对播放器为所欲为了（

比如，想要控制倍数就只需要一句 js 搞定：

`r.controller.speedList.value(1)`

</details>

本来想着可能有暴露出来的接口，事实上也确实是有，但就是没有控制倍数的接口，能完全控制播放器的这个对象并没有被暴露出来。

直接调用接口去解决问题是不行了，而且通过读源码我还发现调整倍数除了播放器的界面上元素的更新，事实上会影响弹幕的播放速度：

```javascript
      $(i).on('ratechange', (function (e) {
        a.danmaku && a.danmaku.option && a.danmaku.option('videospeed', i.playbackRate || 1),
        a.advDanmaku && a.advDanmaku.option && a.advDanmaku.option('videospeed', i.playbackRate || 1)
      })),
```

想要修改 B 站存储的播放倍数也不可能，因为默认情况下这个值被放在了 sessionStorage 中，不说其它的问题，如果修改了这个值，理论上刷新页面也是会生效的，但是，我不明白 B 站程序员为什么要在 unload 的时候把倍数重置：

```javascript
      this.player.window.onunload = function () {
        u.setSession('video_status', 'videospeed', ''),
        u.setSession('video_status', 'videomirror', ''),
        u.setSession('video_status', 'videosize', '')
      },
```

综上所述，目前想要完美实现默认播放倍数设定的话，必须解决以下问题：

-  [ ] 如何控制弹幕播放速度
-  [ ] 更新播放器界面元素

看起来事情没想象中那么简单。。。最困难的就是控制弹幕播放速度，这个应该也是没有暴露接口的，好在弹幕出现是按照时间点来的，就算弹幕滚动的速度不变，强行调整视频播放速度也没有太大影响。

理论上来说，可以通过触发倍数菜单上绑定的点击事件来完成界面更新和弹幕播放速度控制，但是通过快捷键设置的倍数将突破加速上限（2.0x），而且也是直接控制的 playbackRate，因此下面这部分代码也得更新：

```javascript
// src/video/keymap.ts
    const playbackRates = [0.5, 0.75, 1.0, 1.25, 1.5, 2.0, 2.5, 3.0]
      let preventDefault = true
      if (key === '>' || key === 'ArrowUp'.toLowerCase()) { // 天堂制造
        video.playbackRate = playbackRates.find(it => it > video.playbackRate) || playbackRates[playbackRates.length - 1]
        showPlaybackTip(video.playbackRate)
      } else if (key === '<' || key === 'ArrowDown'.toLowerCase()) { // 时间减速
        video.playbackRate = [...playbackRates].reverse().find(it => it < video.playbackRate) || playbackRates[0]
        showPlaybackTip(video.playbackRate)
      } else if (key === '?') { // 重置速度
        video.playbackRate = 1
        showPlaybackTip(video.playbackRate)
      } else if (key === 'w') { // 稍后再看
        const watchlater = dq('.video-toolbar .ops .watchlater, .more-ops-list .ops-watch-later, .video-toolbar-module .see-later-box') as HTMLElement
        if (watchlater !== null) {
          watchlater.click()
```

**Reply by the1812** [reply]
辛苦了 弹幕速度可以先放一放 默认是不同步倍速的
<img height="300" src="https://user-images.githubusercontent.com/26504152/102289456-afebf000-3f79-11eb-994e-aaf27e207e82.png">

更新界面元素可以用 `MutationObserver` 等在面板出现时同步过去

~~我也很烦同步这些界面元素, 所以当时直接就不做了~~

**Reply by LonelySteve** [reply]
我修改了下原评论，理论上可以通过触发点击事件完成界面元素更新，原版范围内的倍数都没什么问题。
